nntrainer/src/databuffer.cpp

```diff
@@ -220,147 +220,87 @@ int DataBuffer::clear() {
   return status;
 }
 
-bool DataBuffer::getDataFromBuffer(BufferType type, vec_4d &outVec,
-                                   vec_4d &outLabel) {
-  unsigned int J, i, j, k, L, l;
-  unsigned int width = input_dim.width();
-  unsigned int height = input_dim.height();
-  unsigned int channel = input_dim.channel();
+bool DataBuffer::getDataFromBuffer(BufferType type,
+                                   std::shared_ptr<float> &out_vec,
+                                   std::shared_ptr<float> &out_label) {
 
-  switch (type) {
-  case BUF_TRAIN: {
-    std::vector<int> list;
-    while (true) {
-      std::unique_lock<std::mutex> ultrain(readyTrainData);
-      cv_train.wait(ultrain, [this]() -> bool { return trainReadyFlag; });
-      if (trainReadyFlag == DATA_ERROR || trainReadyFlag == DATA_END) {
-        if (train_data.size() < batch_size)
-          return false;
-        else
-          break;
-      }
-      if (trainReadyFlag == DATA_READY && train_data.size() >= batch_size) {
-        break;
-      }
-    }
-
-    for (k = 0; k < batch_size; ++k) {
-      std::vector<std::vector<std::vector<float>>> v_channel;
-      for (l = 0; l < channel; ++l) {
-        L = l * width * height;
-        std::vector<std::vector<float>> v_height;
-        for (j = 0; j < height; ++j) {
-          J = L + j * width;
-          std::vector<float> v_width;
-          for (i = 0; i < width; ++i) {
-            v_width.push_back(train_data[k][J + i]);
-          }
-          v_height.push_back(v_width);
-        }
-        v_channel.push_back(v_height);
-      }
-      outVec.push_back(v_channel);
-      outLabel.push_back({{train_data_label[k]}});
-    }
+  using QueueType = std::vector<std::vector<float>>;
 
-    data_lock.lock();
-
-    train_data.erase(train_data.begin(), train_data.begin() + batch_size);
-    train_data_label.erase(train_data_label.begin(),
-                           train_data_label.begin() + batch_size);
-    cur_train_bufsize -= batch_size;
-  } break;
-  case BUF_VAL: {
-    std::vector<int> list;
+  auto wait_for_data_fill = [](std::mutex &ready_mutex,
+                               std::condition_variable &cv, DataStatus &flag,
+                               const unsigned int batch_size,
+                               QueueType &queue) {
     while (true) {
-      std::unique_lock<std::mutex> ulval(readyValData);
-      cv_val.wait(ulval, [this]() -> bool { return valReadyFlag; });
-      if (valReadyFlag == DATA_ERROR || valReadyFlag == DATA_END) {
-        if (val_data.size() < batch_size)
-          return false;
-        else
-          break;
-      }
-      if (valReadyFlag == DATA_READY && val_data.size() >= batch_size) {
-        break;
-      }
-    }
+      std::unique_lock<std::mutex> ul(ready_mutex);
+      cv.wait(ul, [&]() -> bool { return flag; });
```

Comment:
condition_variables have spurios wakes - https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-wait
So, there can be spurious wakeups, and this will just throw for it - it must  NOT.